src/lib.rs

```diff
@@ -942,56 +928,70 @@ impl<'a> PtpCamera<'a> {
-                };
-                let timespec = time::get_time();
-                trace!("Read Data  [{}:{:09}] - length:{:?} remaining:{:?}",
-                       timespec.sec,
-                       timespec.nsec,
-                       current_len,
-                       remaining_buf.len());
-                let len = try!(self.handle.read_bulk(self.ep_in, remaining_buf, timeout));
-                unsafe {
-                    self.buf.set_len(current_len + len);
-                }
-                // debug!("new buf len [{:?}] into {:?}", self.buf.len(), remaining_buf.len());
-                if len < remaining_buf.len() {
-                    break;
+            let (container, payload) = try!(self.read_txn_phase());
+            if container.belongs_to(tid) {
+                match container.kind {
+                    PtpContainerType::Data => {
+                        data_phase_payload = payload;
+                    },
+                    PtpContainerType::Response => {
+                        if container.code != StandardResponseCode::Ok {
+                            return Err(Error::Response(container.code));
+                        }
+                        return Ok(data_phase_payload);
+                    },
+                    _ => {}
                 }
+            } else {
+                debug!("mismatched txnid {}, expecting {}", container.tid, tid);
             }
+        }
+    }
 
-            let (mtype, mut msg) = try!(PtpTransaction::parse(&self.buf));
+    // helper for command() above, retrieve container info and payload for the current phase
+    fn read_txn_phase(&mut self) -> Result<(PtpContainerInfo, Vec<u8>), Error> {
+        // large enough to fetch objects other than media in a single read
+        let mut buf = Vec::with_capacity(100 * 1024);
+        let timeout = Duration::from_secs(2);
 
-            if mtype == PtpContainerType::Data && msg.is_response(&transaction) {
-                data = Some(msg.data);
-            } else if mtype == PtpContainerType::Response && msg.is_response(&transaction) {
-                if let Some(data) = data {
-                    msg.data = data;
-                }
-                
-                if msg.code != StandardResponseCode::Ok {
-                    return Err(Error::Response(msg.code));
-                }
-                
-                return Ok(msg.data);
+        let bufslice = unsafe {
+            slice::from_raw_parts_mut(buf.as_mut_ptr(), buf.capacity())
+        };
+        let n = try!(self.handle.read_bulk(self.ep_in, bufslice, timeout));
+        unsafe { buf.set_len(n); }
+
+        let cinfo = try!(PtpContainerInfo::parse(&buf[..]));
+        trace!("container {:?}, bulk rx {}", cinfo, n);
+
+        // response didn't fit into our original buf? resize and read the rest
+        // should ideally require a single read, but loop to handle short reads.
```

Comment:
looks like a short packet should only be sent to indicate the end of a command phase, so could otherwise treat it as an error. will update.